### ISSUE TITLE
'<!> Class "RadioMedium" not found' when running artery from root directory with symlink(s) in its path

### ISSUE THREAD

**Issue by m-wegner** [riebl/artery#28]
To optimize for single-thread performance, we set up a new simulation server based on a Intel Xeon W-2125 CPU with Ubuntu 16.04.5 LTS.

For an MWE, we cloned the artery repository on branch master and built artery following the README. We get no compile or linker errors throughout veins, vanetza, inet and artery.

When trying to run the included artery example in OMNeT++'s Cmdenv with the following script:
```shell
cd /path/to/artery/scenarios/artery/ && \
        /path/to/contrib/omnetpp-5.4/bin/opp_run_release \
        -n /path/to/artery/src/artery:/path/to/artery/src/traci:/path/to/artery/extern/veins/examples/veins:/path/to/artery/extern/veins/src/veins:/path/to/artery/extern/inet/src:/path/to/artery/extern/inet/examples:/path/to/artery/extern/inet/tutorials:/path/to/artery/extern/inet/showcases \
        -l /path/to/artery/build/src/artery/envmod/libartery_envmod.so \
        -l /path/to/artery/build/src/artery/envmod/libartery_envmod.so \
        -l /path/to/artery/build/src/artery/storyboard/libartery_storyboard.so \
        -l /path/to/artery/extern/inet/out/gcc-release/src/libINET.so \
        -l /path/to/artery/extern/veins/out/gcc-release/src/libveins.so \
        -l /path/to/artery/build/src/artery/libartery_core.so \
        omnetpp.ini \
        -u Cmdenv -c envmod -r 0
```

We get the following error:
```
OMNeT++ Discrete Event Simulation  (C) 1992-2017 Andras Varga, OpenSim Ltd.
Version: 5.4, build: 180611-572210e, edition: Academic Public License -- NOT FOR COMMERCIAL USE
See the license for distribution terms and warranty disclaimer

Setting up Cmdenv...

Loading NED files from /ibr/projects/artery-lte/artery/src/artery:  53
Loading NED files from /ibr/projects/artery-lte/artery/src/traci:  15
Loading NED files from /ibr/projects/artery-lte/artery/extern/veins/examples/veins:  1
Loading NED files from /ibr/projects/artery-lte/artery/extern/veins/src/veins:  38
Loading NED files from /ibr/projects/artery-lte/artery/extern/inet/src:  666
Loading NED files from /ibr/projects/artery-lte/artery/extern/inet/examples:  174
Loading NED files from /ibr/projects/artery-lte/artery/extern/inet/tutorials:  19
Loading NED files from /ibr/projects/artery-lte/artery/extern/inet/showcases:  33

Preparing for running configuration envmod, run #0...
Scenario: $preseletion=1, $repetition=0
Assigned runID=envmod-0-20180820-11:38:10-42030
Setting up network "artery.envmod.World"...

[Warning: Program was compiled without debug info, you might not be able to debug.]

RUNTIME ERROR. A cRuntimeError exception is about to be thrown, and you
requested (by setting debug-on-errors=true in the ini file) that errors
abort execution and break into the debugger.

You should now probably be running the simulation under gdb or another
debugger. The simulation kernel will now raise a SIGINT signal which will
get you into the debugger. If you are not running under a debugger, you can
still use the core dump for post-mortem debugging. Once in the debugger,
view the call stack (in gdb: "bt" command) to see the context of the
runtime error.

<!> Class "RadioMedium" not found -- perhaps its code was not linked in, or the class wasn't registered with Register_Class(), or in the case of modules and channels, with Define_Module()/Define_Channel() -- in module (omnetpp::cModule) World (id=1), during network setup

TRAPPING on the exception above, due to a debug-on-errors=true configuration option. Is your debugger ready?
```

(Probably irrelevant) gdb stack trace at this point:
```
#0  0x00007ffff7509060 in omnetpp::cRuntimeError::notifyEnvir() () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#1  0x00007ffff7509221 in omnetpp::cRuntimeError::cRuntimeError(char const*, ...) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#2  0x00007ffff74cfd7e in omnetpp::cObjectFactory::get(char const*, char const*, bool) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#3  0x00007ffff74cfdc2 in omnetpp::cObjectFactory::createOne(char const*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#4  0x00007ffff74d8078 in omnetpp::cModuleType::instantiateModuleClass(char const*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#5  0x00007ffff74d96a2 in omnetpp::cModuleType::create(char const*, omnetpp::cModule*, int, int) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#6  0x00007ffff7616e9e in omnetpp::cNedNetworkBuilder::addSubmodule(omnetpp::cModule*, omnetpp::nedxml::SubmoduleElement*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#7  0x00007ffff761731e in omnetpp::cNedNetworkBuilder::addSubmodulesAndConnections(omnetpp::cModule*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#8  0x00007ffff7617413 in omnetpp::cNedNetworkBuilder::buildRecursively(omnetpp::cModule*, omnetpp::cNedDeclaration*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#9  0x00007ffff761751a in omnetpp::cNedNetworkBuilder::buildInside(omnetpp::cModule*, omnetpp::cNedDeclaration*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#10 0x00007ffff7610302 in omnetpp::cDynamicModuleType::buildInside(omnetpp::cModule*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#11 0x00007ffff752d5a2 in omnetpp::cModule::buildInside() () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#12 0x00007ffff755aa4c in omnetpp::cSimulation::setupNetwork(omnetpp::cModuleType*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppsim.so
#13 0x00007ffff79463d1 in omnetpp::envir::EnvirBase::setupNetwork(omnetpp::cModuleType*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppenvir.so
#14 0x00007ffff7bcf9ec in omnetpp::cmdenv::Cmdenv::doRun() () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppcmdenv.so
#15 0x00007ffff794366d in omnetpp::envir::EnvirBase::run() () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppenvir.so
#16 0x00007ffff7948a2c in omnetpp::envir::EnvirBase::run(int, char**, omnetpp::cConfiguration*) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppenvir.so
#17 0x00007ffff793e565 in omnetpp::envir::setupUserInterface(int, char**) () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppenvir.so
#18 0x00007ffff793fe10 in evMain () from /misc/ibr/projects/artery-lte/contrib/omnetpp-5.4/lib/liboppenvir.so
#19 0x00007ffff5aea830 in __libc_start_main (main=0x400670 <main>, argc=22, argv=0x7fffffffd508, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd4f8) at ../csu/libc-start.c:291
```

Tested with OMNeT++ versions 5.4 and 5.4.1. This only happens with this server. Other hardware (and/or distributions, namely ArchLinux and Gentoo) is fine.

So, we're kindly asking for any ideas on what's happening here? Some sort of race condition when setting up modules?

**Comment by hagau** [comment]
trace with debug symbols
```
#0  omnetpp::cRuntimeError::notifyEnvir (this=0x15144c0) at cexception.cc:260
#1  0x00007ffff7378a7c in omnetpp::cRuntimeError::cRuntimeError (this=0x15144c0, 
    msgformat=0x7ffff74d19f8 "Class \"%s\" not found -- perhaps its code was not linked in, or the class wasn't registered with Register_Class(), or in the case of modules and channels, with Define_Module()/Define_Channel()") at cexception.cc:234
#2  0x00007ffff73261be in omnetpp::cObjectFactory::get (className=0x14b4668 "RadioMedium", contextNamespace=0x0, fallbackToOmnetpp=true) at cobjectfactory.cc:52
#3  0x00007ffff73262c0 in omnetpp::cObjectFactory::createOne (classname=0x14b4668 "RadioMedium") at cobjectfactory.cc:65
#4  0x00007ffff733683f in omnetpp::cModuleType::instantiateModuleClass (this=0x14b4a60, className=0x14b4668 "RadioMedium") at ccomponenttype.cc:357
#5  0x00007ffff74af393 in omnetpp::cDynamicModuleType::createModuleObject (this=0x14b4a60) at netbuilder/cdynamicmoduletype.cc:66
#6  0x00007ffff73364c7 in omnetpp::cModuleType::create (this=0x14b4a60, moduleName=0x7d27b0 "radioMedium", parentModule=0x1513340, vectorSize=-1, index=0) at ccomponenttype.cc:293
#7  0x00007ffff733631f in omnetpp::cModuleType::create (this=0x14b4a60, moduleName=0x7d27b0 "radioMedium", parentModule=0x1513340) at ccomponenttype.cc:265
#8  0x00007ffff74b3f00 in omnetpp::cNedNetworkBuilder::addSubmodule (this=0x7fffffffd8f0, compoundModule=0x1513340, submod=0x7d2730) at netbuilder/cnednetworkbuilder.cc:763
#9  0x00007ffff74b2059 in omnetpp::cNedNetworkBuilder::addSubmodulesAndConnections (this=0x7fffffffd8f0, modp=0x1513340) at netbuilder/cnednetworkbuilder.cc:474
#10 0x00007ffff74b1ff1 in omnetpp::cNedNetworkBuilder::buildRecursively (this=0x7fffffffd8f0, modp=0x1513340, decl=0x13489e0) at netbuilder/cnednetworkbuilder.cc:466
#11 0x00007ffff74b1e66 in omnetpp::cNedNetworkBuilder::buildInside (this=0x7fffffffd8f0, modp=0x1513340, decl=0x13489e0) at netbuilder/cnednetworkbuilder.cc:443
#12 0x00007ffff74af5ab in omnetpp::cDynamicModuleType::buildInside (this=0x1349090, module=0x1513340) at netbuilder/cdynamicmoduletype.cc:89
#13 0x00007ffff73a1975 in omnetpp::cModule::doBuildInside (this=0x1513340) at cmodule.cc:1187
#14 0x00007ffff73a18b0 in omnetpp::cModule::buildInside (this=0x1513340) at cmodule.cc:1176
#15 0x00007ffff73d8c12 in omnetpp::cSimulation::setupNetwork (this=0x67c480, network=0x1349090) at csimulation.cc:388
#16 0x00007ffff78f44c9 in omnetpp::envir::EnvirBase::setupNetwork (this=0x78b5c0, network=0x1349090) at envirbase.cc:766
#17 0x00007ffff7bcd9b6 in omnetpp::cmdenv::Cmdenv::doRun (this=0x78b5c0) at cmdenv.cc:243
#18 0x00007ffff78f4374 in omnetpp::envir::EnvirBase::run (this=0x78b5c0) at envirbase.cc:742
#19 0x00007ffff78f1dce in omnetpp::envir::EnvirBase::run (this=0x78b5c0, argc=16, argv=0x7fffffffe208, configobject=0x683850) at envirbase.cc:354
#20 0x00007ffff78ea1ee in omnetpp::envir::setupUserInterface (argc=16, argv=0x7fffffffe208) at startup.cc:259
#21 0x00007ffff78eb6ee in omnetpp::envir::evMain (argc=16, argv=0x7fffffffe208) at evmain.cc:33
#22 0x0000000000400796 in main (argc=16, argv=0x7fffffffe208) at main.cc:31
```

**Comment by hagau** [comment]
I'd like to note that there is a `inet::physicallayer::RadioMedium` in the global `classes` registry before the exception.
A `classes.getInstance()->lookup("RadioMedium", "inet::physicallayer", true)` returns the desired `omnetpp::cObjectFactory`.

**Comment by hagau** [comment]
As reference, on my system `omnetpp::cModuleType::instantiateModuleClass` gets called with `className`=`inet::physicallayer::RadioMedium`.  
The `className` comes from `implClassName` of the `omnetpp::cNedDeclaration` of `inet.physicallayer.ieee80211.packetlevel.Ieee80211ScalarRadioMedium`, which should be `inet::physicallayer::RadioMedium`, but is `RadioMedium` on the afflicted system.  

So this is some sort of parsing issue in the NED xml parser maybe?

**Comment by hagau** [comment]
Further observations:
- fixing the namespace of the `className` argument in gdb works
 - after `RadioMedium`, some other modules are added correctly until `ConstantSpeedPropagation`, after which every `className` is missing its proper namespace.

For some reason, the `implClassName` of the `omnetpp::cNedDeclaration` for some classes is not namespaced properly.

**Comment by torokati44** [comment]
We came across an issue recently that likely shares a root cause with this one. See the fix for that here: https://github.com/inet-framework/inet/commit/bfa37126b7617c97aba4f7917ccf3c47c502d5ef

In a scenario where a module type in a project extends a module type in another project (say, Artery and INET), but only on the NED level, without its own custom C++ class, the class name lookup can fail.
In this case, the `@class` property of the supertype (which is in a namespace) is inherited by the subtype - in a different namespace.

If the supertype does not specify the namespace of the class inline (like `@class(inet::physicallayer::RadioMedium)`), but instead relies on the preceding standalone `namespace` statements (like INET does), then the class will be searched for in the namespace of the subtype (like `artery`).

To work around this, you could add explicit overrides to any such subtype with fully qualified class names: `@class(inet::physicallayer::RadioMedium)`

However, we are not aware of anything that would make this hardware- or OS-dependent. Are you absolutely sure that the exact same revisions of all the numerous components are running in the working and non-working cases? It might just be that this corner case was only introduced in a recent revision of one of the projects.

If my description was not clear, feel free to ask for more details.

**Comment by m-wegner** [comment]
Thank you for your very thorough explanation, @torokati44! It reads to me as if this could be the root cause, but this has to be checked, of course. We will see to try the mentioned possibilities for a fix. Thanks!

As to your question regarding hardware- or OS-dependencies: I, too, was a bit reluctant to attribute this error that we are seeing to the hardware or OS, as I am not aware of, e.g., any parallelism that would allow for such a dependency.

However, we were and are testing this with the exact same revision (i.e., branch/commit hash) of artery (and subsequently its bundled INET revision) as well as the exact same OMNeT++ version(s) (as described we tested 5.4 and 5.4.1). We also eliminated different compilers as possible cause. Yet, we still have the simulations working on two developer machines which after checking all these things only differ in Linux distribution and hardware.

**Comment by hagau** [comment]
@torokati44 Thank you, that makes sense to me. 

Hm, we do not subtype `inet::physicallayer::RadioMedium` here:
https://github.com/riebl/artery/blob/4a5b17c56d9897aa408f114dab5e9d4edd8c2ef7/src/artery/inet/World.ned#L22

**Comment by torokati44** [comment]
And by "INET source root" I mean the "src" directory, not the git repo root, sorry.

**Comment by hagau** [comment]
- NEDPATH contains INET root
- `src/inet/package.ned` does contain `@namespace(inet)`
- there are `@namespace(...)` directives in `src/inet/` but they all make sense, e.g.  
`src/inet/physicallayer/package.ned` has `@namespace(inet::physicallayer)`

**Comment by torokati44** [comment]
Just a random idea: What if you change the NEDPATH to contain either the `src` directory of INET, or `src/inet`? Does it make a difference having one or the other?

**Comment by hagau** [comment]
`src` or `src/inet` make no difference, still the same issue.

**Comment by hagau** [comment]
I'm using the same script to launch the simulation on both systems, the afflicted one and my development system (where everything works perfectly)

**Comment by torokati44** [comment]
We'll look into it a bit more tomorrow.

**Comment by hagau** [comment]
@torokati44 Thanks, and thank you for the quick reply!

**Comment by torokati44** [comment]
We did not forget about this, we just have to prepare for the annual [Community Summit](https://summit.omnetpp.org/2018/).
We have to postpone this for about a week, sorry.

**Comment by m-wegner** [comment]
So, I've done some further investigation regarding my assumption that this has to be related to a root cause within the underlying system and dependencies of the simulation framework, here are my findings so far:

- An upgrade of the affected machine to Ubuntu 18.04 did not solve this.
- We booted an ArchLinux live system on the affected simulation node and within this ArchLinux we were successfully able to run the artery example and our own implementation and scenarios, without running in this issue.
- A (rather brief) look at the ned parser in OMNeT++ shows a dependency to bison, probably flex, m4 and libxml2.
- From these only bison and libxml2 differed in their versions between the ArchLinux and Ubuntu environments.
- Yet, compiling both in their Arch versions on Ubuntu 18.04 (bison in version 3.0.5 and libxml2 in version 2.9.8) and forcing a recompilation of OMNeT++ and whole artery code (incl. INET, etc.) with these exact versions did not yield any changes, i.e., bug is still there.
- We were able to reproduce this bug on another machine with Ubuntu 16.04 and a Intel Xeon E5520.

So, in summary:
- it seems that this bug is not hardware-dependent (at least not specifically to the initally mentioned CPU). I adjusted the title of this issue accordingly.
- we are able to reproduce this with Ubuntu 16.04 and 18.04. ArchLinux and Gentoo distributions seem not to be affected. So, we assume this bug to be somewhere in the diff of artery's and OMNeT++'s/INET's dependencies within these distributions. Yet, I'd expect more users of artery (or even INET/OMNeT++) to be affected by this.

**Comment by riebl** [comment]
@m-wegner Can you try to build Artery on Ubuntu with another compiler, i.e. one that has not been "improved" by Ubuntu? I had already quite some fun with Ubuntu's specific compiler tuning…

**Comment by m-wegner** [comment]
@riebl Yeah, I think, we could try upstream clang, I've done this before with Ubuntu. As I have to run this by our admin, this will have to wait till tomorrow, though (I'd like to just use LLVM's PPA, and not set clang up myself). :wink: But thanks for the suggestion! We'll try that.

**Comment by torokati44** [comment]
Could you try to print the `mangledName` parameter of `opp_demangle_typename` (in `src/sim/util.cc`), as it is passed in?
Also, the return value? - This might be the easiest to print from `opp_typename` (as this is the only place where `opp_demangle_typename` is called from).
Is there any difference in these between the affected and unaffected setups?

**Comment by Venturina** [comment]
@m-wegner Just as a note: I'm using Ubuntu 18.04 and I have tried to reproduce your error, but I'm not experiencing any error when running the Artery example scenario.
Do I have to change anything in the scenario to reproduce it?

**Comment by hagau** [comment]
@Venturina No, this should be reproducible with a clean checkout.

**Comment by m-wegner** [comment]
> Do I have to change anything in the scenario to reproduce it?

No, we used plain OMNeT++ 5.4 or 5.4.1, respectively (compiled without osg, osgearth and qtenv/tkenv support). Artery built according to README and we launched the scenario as stated in the initial comment in this issue: We just changed the command line to run the Cmdenv of OMNeT++ and chose the `envmod` config section of `scenario/artery/omnetpp.ini`.

**Comment by m-wegner** [comment]
@torokati44 Thanks for your suggestion, we'll try that as well (probably also by tomorrow) and will report back with the results.

**Comment by m-wegner** [comment]
@Venturina Forgot to say: Thanks for joining and confirming a working setup with Ubuntu 18.04! By now, I'm really intrigued what the core issue here is (and how bizarre it will turn out to be, if we really get to the bottom of it that is, anyways).

**Comment by hagau** [comment]
On both systems I see
```
mangledName:  N4inet13physicallayer11RadioMediumE
opp_demangle_typename ret:  inet::physicallayer::RadioMedium
```

but on the affected system the last two calls to `opp_demangle_typename` before the exception are:
```
mangledName:  N7omnetpp7cModuleE
opp_demangle_typename ret:  omnetpp::cModule

mangledName:  N7omnetpp7cModuleE
opp_demangle_typename ret:  omnetpp::cModule
```

**Comment by torokati44** [comment]
Thanks!

Could you now try replacing the `NedTypeInfo::getPackageProperty` method in `src/nedxml/nedtypeinfo.cc` with this annotated version?

```c++
std::string NedTypeInfo::getPackageProperty(const char *propertyName) const
{
    std::cout << "Finding package property: " << propertyName << " for type " << this->getFullName() << std::endl;

    // find first such property in the current file, then in package.ned of this package and parent packages
    for (NedFileElement *nedFile = (NedFileElement *)getTree()->getParentWithTag(NED_NED_FILE);
         nedFile != nullptr;
         nedFile = getResolver()->getParentPackageNedFile(nedFile))
    {
        std::cout << "Visiting NED file: " << nedFile->getFilename() << std::endl;
        const char *propertyValue = ASTNodeUtil::getLocalStringProperty(nedFile, propertyName);
        if (propertyValue) {
            std::cout << "returning: " << propertyValue << std::endl;
            return propertyValue;
        }
    }
    std::cout << "returning: EMPTY" << std::endl;
    return "";
}
```

... and posting what it prints in both cases?

**Comment by hagau** [comment]
On the working setup:
```
Finding package property: namespace for type inet.physicallayer.common.packetlevel.RadioMedium
Visiting NED file: /{...}/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
Visiting NED file: /{...}/artery/extern/inet/src/inet/physicallayer/package.ned
returning: inet::physicallayer
```
on the affected system:
```
Finding package property: namespace for type inet.physicallayer.common.packetlevel.RadioMedium                                                                                  
Visiting NED file: /{...}/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned                                             
returning: EMPTY
```

**Comment by hagau** [comment]
It always returns "" from `NedTypeInfo::getPackageProperty` on the affected system and never finds the parent `package.ned`...

**Comment by torokati44** [comment]
And I presume there _is_ a `{...}/artery/extern/inet/src/inet/physicallayer/package.ned` file on both systems, with identical contents...

Are the filesystems the same type? Are the file permissions different? Any uppercase/lowercase differences, or case sensitivity differences? (Probably all of them are the same, but just in case...)

How about if you add the `cout` line to `NedResourceCache::addFile` in this place:

```c++
    NedFileMap::iterator it = files.find(key);
    if (it != files.end())
        return false;  // already added

    std::cout << "added file " << fname << " using key " << key << std::endl; // add this line

    files[key] = node;
```

**Comment by Venturina** [comment]
On my working Ubuntu 18.04 I'm also using omnet 5.4.1 without osg and osgearth and tkenv (so the build differs only at the qtenv option). I'm also starting the scenario using your script.

I'm not sure if it is helping you, but I have tried the same steps to verify my setup.

Changing the `NedTypeInfo::getPackageProperty`outputs the same as yours on the working setups.

```
Finding package property: namespace for type inet.physicallayer.common.packetlevel.RadioMedium
Visiting NED file: /path/to/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
Visiting NED file: /path/to/artery/extern/inet/src/inet/physicallayer/package.ned
returning: inet::physicallayer
```

Adding the above mentioned `cout` line prints the following regarding the `RadioMedium.ned`

```
added file path/to/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned using key /path/to/artery/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
```

**Comment by hagau** [comment]
- Yes, the file exist & has the correct spelling and permissions
- Well, on the affected system it's a NFS mounted ZFS. Should not be a problem, I think.
NedResourceCache::addFile  

The addition to `NedResourceCache::addFile` produces:
```
added file {...}/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned using key {...}/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned                                                                                                                
```

**Comment by hagau** [comment]
@Venturina Thanks for verifying!

**Comment by hagau** [comment]
package.ned is also added:
```
added file /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/package.ned using key /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/ph
ysicallayer/package.ned
```

**Comment by torokati44** [comment]
Thank you so much for the quick responses. We are clueless so far. We'll get back to this tomorrow. :)

**Comment by hagau** [comment]
@torokati44 Thank you for your help! :)

**Comment by torokati44** [comment]
If you replace these two functions in `nedresourcecache.cc` with these annotated versions:
https://gist.github.com/torokati44/b644f08758d5f4cff3dffd3e8f7ad9f0

Where do they diverge?

**Comment by hagau** [comment]
On the affected system:
```
Finding package property: namespace for type inet.physicallayer.common.packetlevel.RadioMedium              
Visiting NED file: /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned                                              
getParentPackage for /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned                                            
topDir:
returning: EMPTY
```

**Comment by hagau** [comment]
I butterfingered the `make` on my working setup, so that will take a bit of time, since recompile :)

But I noticed something on the affected system:
```
8158:Finding package property: namespace for type inet.physicallayer.common.packetlevel.RadioMedium
8159:Visiting NED file: /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
8160:getParentPackage for /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
8161-absPath: /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel
8162-folderName: /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel
8163-folderPackage: /ibr/projects/artery-lte/artery_hagau/extern/inet/src
8164-folderPackage: /ibr/projects/artery-lte/artery_hagau/extern/veins/examples/veins
8165-folderPackage: /ibr/projects/artery-lte/artery_hagau/extern/veins/src/veins
8166-folderPackage: /ibr/projects/artery-lte/artery_hagau/src/artery
8167-folderPackage: /ibr/projects/artery-lte/artery_hagau/src/traci
8168-topDir:
8169-returning: EMPTY
```

**Comment by hagau** [comment]
On my working setup, as expected:
```
Finding package property: namespace for type inet.physicallayer.common.packetlevel.RadioMedium
Visiting NED file: /home/artery/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
getParentPackage for /home/artery/artery/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
topDir: /home/artery/artery/extern/inet/src
```

**Comment by hagau** [comment]
So, the culprit is using the `/ibr` symlink.

```
13548:Finding package property: namespace for type inet.physicallayer.common.packetlevel.RadioMedium
13549:Visiting NED file: /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
13550:getParentPackage for /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
13551-absPath: /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel
13552-folderName: /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel
13553-folderPackage.f: /ibr/projects/artery-lte/artery_hagau/extern/inet/src
13554-folderPackage.s:
13555-isPathPrefixOf: comparing: /ibr/projects/artery-lte/artery_hagau/extern/inet/src  and   /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/pac
ketlevel
13556-isPathPrefixOf:2 not prefix
```
It works if I use the full path.

**Comment by hagau** [comment]
@torokati44 Thanks for the pointer to look at the filesystem, I would not have checked again otherwise :)

**Comment by torokati44** [comment]
So, at some point, the symlink gets resolved, and this causes some mismatches down the line? Between the unresolved and resolved paths?

**Comment by hagau** [comment]
No, the symlink not getting resolved is the problem. For some reason, `folderPackages` in `nedresourcecache.cc` gets initialized with the unresolved path. I'm still investigating the root cause...

**Comment by torokati44** [comment]
Well, there are a couple `getcwd` calls in `toAbsolutePath`. I'm not sure if that consistently returns either a resolved or an unresolved path.

**Comment by hagau** [comment]
Hm, the `getcwd` isn't reached in this case, since the path starts with a `/`.

**Comment by torokati44** [comment]
But then where do any paths that don't go through the symlink come from?

**Comment by hagau** [comment]
I don't know :)
`$OMNETPP_HOME` is also set to `/ibr/...`, and this works when I set the `$NEDPATH` to the resolved path.

**Comment by hagau** [comment]
In `NedResourceCache::loadNedSourceFolder`, should not the `canonicalFolderName` be the resolved pathname? I've got `/ibr/..` on the affected system.

**Comment by hagau** [comment]
@torokati44 Thank you very much for your help! We'll finish this next week :)

**Comment by torokati44** [comment]
Well, isn't the problem that the symlink is resolved somewhere, even though it shouldn't be?

I mean, the goal is consistency, but we should clarify first whether we should have resolved or unresolved path names everywhere. My vote is on unresolved, since what symlinks you have on your system on the path leading to OMNeT++ should not be relevant - especially not where they point.

And I don't think OMNeT++ at the moment manually tries to resolve symlinks, neither do I think it should. (It only converts relative paths to absolute ones, and deals with `.` and `..`.)
The only thing I can think of is that some filesystem/path manipulation function we use resolves them anyway, and maybe we should prevent that?

**Comment by hagau** [comment]
You are right, `getcwd` is the offending call
```
5264:getting file: RadioMedium.ned NOT FOUND                                                                                                                                         
5265:toAbsolutePath:pathname RadioMedium.ned                                                                                                                                         
5266:toAbsolutePath:ret /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned                                             
5267:added file /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned using key /misc/ibr/projects/artery-lte/artery_hagau/extern/inet/src/inet/physicallayer/common/packetlevel/RadioMedium.ned
```
In this case the `getcwd` in `toAbsolutePath` is called and expands the symlink.
(The man page doesn't mention it, but http://pubs.opengroup.org/onlinepubs/9699919799/functions/getcwd.html does)

**Comment by m-wegner** [comment]
Simulations work fine when running everything from the canonical/resolved path. So, at least for our case we have a workaround, though a proper fix is probably needed in OMNeT++. Big thanks again to everyone involved!

**Comment by riebl** [comment]
Needs to be solved in OMNeT++, see bug tracker entry. Closing.

**Comment by rima-boughariou** [comment]
I have not been able to solve this error for a long time, I need help.
Class "inet::physicallayer::RadioMedium" not found -- perhaps its code was not linked in, or the class wasn't registered with Register_Class(), or in the case of modules and channels, with Define_Module()/Define_Channel() -- in module (omnetpp::cModule) Scenario (id=1), during network setup

**Comment by riebl** [comment]
Please create a new issue ticket for your problem instead of hijacking issue tickets closed years ago.

By the way, the actual reason why your setup fails is usually printed before the *Class "inet::physicallayer::RadioMedium" not found" error message. Look out for "undefined symbol" messages.
